File: AndroidGameLibrary/src/cc/lib/android/SpinnerTask.java
package cc.lib.android;

import android.app.Activity;
import android.app.AlertDialog;
import android.app.Application;
import android.app.ProgressDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.os.AsyncTask;
import android.os.Bundle;

import org.jetbrains.annotations.NotNull;

import java.lang.ref.WeakReference;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

public abstract class SpinnerTask<T> extends AsyncTask<T, Integer, Object>
        implements Application.ActivityLifecycleCallbacks {

    private ProgressDialog dialog = null;
    private final WeakReference<Activity> context;
    private Object result = null;

    public SpinnerTask(Activity context) {
        this.context = new WeakReference<>(context);
        if (context instanceof Activity) {
            // we want to handle the activity pausing so we can cancel ourselves
            context.getApplication().registerActivityLifecycleCallbacks(this);
        }
    }

    public void setProgressMessage(String message) {
        dialog.setMessage(message);
    }

    public <T extends Context> T getContext() {
        return (T) context.get();
    }

    @Override
    protected final void onCancelled(Object o) {
        super.onCancelled(o);
    }

    @Override
    protected final void onCancelled() {
        if (dialog != null)
            dialog.dismiss();
        onCompleted();
    }

    @Override
    protected final void onPostExecute(Object o) {
        if (dialog != null)
            dialog.dismiss();
        if (!isCancelled()) {
            if (o != null && o instanceof Exception) {
                Exception e = (Exception) o;
                onError(e);
            } else {
                onSuccess();
            }
            onCompleted();
        }
    }

    /**
     * Gets called for all situations: SUCCESS, ERROR, CANCELLED
     */
    protected void onCompleted() {}

    protected void onError(@NotNull Exception e) {
        e.printStackTrace();
        Activity c = context.get();
        if (c != null) {
            c.runOnUiThread(() -> {
                new AlertDialog.Builder(c).setTitle("Error").setMessage("An error occured: " + e.getClass().getSimpleName() + "\n" + e.getMessage())
                        .setNegativeButton("Ok", null).show();
            });
        }
    }

    @Override
    protected void onPreExecute() {
        if (!context.get().isFinishing() && !context.get().isDestroyed()) {
            dialog = new ProgressDialog(context.get());
            dialog.setMessage(getProgressMessage());
            dialog.setCancelable(false); // prevents cancel from back button
            dialog.setIndeterminate(true);
            dialog.setCanceledOnTouchOutside(false); // prevents cancel from random touches
            if (isCancellable()) {
                dialog.setButton(DialogInterface.BUTTON_NEGATIVE, context.get().getString(R.string.popup_button_cancel), this::onCancelButtonClicked);
            }
            dialog.show();
        }
    }

    protected String getProgressMessage() {
        return null;
    }

    @Override
    protected void onProgressUpdate(Integer... values) {
        super.onProgressUpdate(values);
    }

    protected abstract void doIt(T ... args) throws Throwable;

    protected void onSuccess() {}

    @Override
    protected final Object doInBackground(T... args) {
        try {
            doIt(args);
        } catch (Throwable e ) {
            return e;
        }
        return null;
    }

    public boolean isCancellable() {
        return true;
    }

    public boolean canInterruptOnCancel() {
        return false;
    }

    protected final void setResult(Object result) {
        this.result = result;
    }

    public final <T> T getResult() {
        return (T)result;
    }

    private void onCancelButtonClicked(DialogInterface dialog, int which) {
        cancel(canInterruptOnCancel());
        onCancelButtonClicked();
        dialog.dismiss();
    }

    protected void onCancelButtonClicked() {}

    @Override
    public final void onActivityCreated(@NonNull Activity activity, @Nullable Bundle savedInstanceState) {

    }

    @Override
    public final void onActivityStarted(@NonNull Activity activity) {

    }

    @Override
    public final void onActivityResumed(@NonNull Activity activity) {

    }

    @Override
    public final void onActivityPaused(@NonNull Activity activity) {

    }

    @Override
    public final void onActivityStopped(@NonNull Activity activity) {
        cancel(true);
    }

    @Override
    public final void onActivitySaveInstanceState(@NonNull Activity activity, @NonNull Bundle outState) {

    }

    @Override
    public final void onActivityDestroyed(@NonNull Activity activity) {

    }
}

File: AndroidGameLibrary/src/cc/lib/android/SpinnerTask.kt
package cc.lib.android

import android.app.Activity
import android.app.AlertDialog
import android.app.Application.ActivityLifecycleCallbacks
import android.app.ProgressDialog
import android.content.DialogInterface
import android.os.Bundle
import cc.lib.utils.takeIfInstance
import cc.lib.utils.weakReference

abstract class SpinnerTask<T>(_context: Activity) : CoroutinesAsyncTask<T, Int, Any>(_context.localClassName),
	ActivityLifecycleCallbacks {
	private var dialog: ProgressDialog? = null
	private val context by weakReference(_context)
	private var result: Any? = null

	init {
		context?.takeIfInstance<Activity>()?.let {
			it.application.registerActivityLifecycleCallbacks(this)
		}
	}

	override fun onCancelled(o: Any?) {
		super.onCancelled(o)
	}

	override fun onCancelled() {
		dialog?.dismiss()
		onCompleted()
	}

	override fun onPostExecute(o: Any?) {
		dialog?.dismiss()
		if (!isCancelled) {
			if (o != null && o is Exception) {
				onError(o)
			} else {
				onSuccess()
			}
			onCompleted()
		}
	}

	/**
	 * Gets called for all situations: SUCCESS, ERROR, CANCELLED
	 */
	protected open fun onCompleted() {}
	protected open fun onError(e: Exception) {
		e.printStackTrace()
		context?.let {
			it.runOnUiThread {
				AlertDialog.Builder(it).setTitle("Error")
					.setMessage(
						"""
		An error occured: ${e.javaClass.simpleName}
		${e.message}
		""".trimIndent()
					)
					.setNegativeButton("Ok", null).show()
			}
		}
	}

	fun postExecute(args: T? = null) {
		context?.runOnUiThread { execute(args) }
	}

	override fun onPreExecute() {
		context?.apply {
			if (!isFinishing && !isDestroyed) {
				dialog = ProgressDialog(this).also {
					it.setMessage(progressMessage)
					it.setCancelable(false) // prevents cancel from back button
					it.isIndeterminate = true
					it.setCanceledOnTouchOutside(false) // prevents cancel from random touches
					if (isCancellable) {
						it.setButton(
							DialogInterface.BUTTON_NEGATIVE,
							getString(R.string.popup_button_cancel)
						) { dialog: DialogInterface, which: Int -> onCancelButtonClicked(dialog, which) }
					}
					it.show()
				}
			}
		}
	}

	protected open var progressMessage: String?
		protected get() = null
		set(message) {
			dialog?.setMessage(message)
		}

	@Throws(Throwable::class)
	protected abstract suspend fun doIt(args: T?)
	protected open fun onSuccess() {}
	override suspend fun doInBackground(args: T?): Any? {
		try {
			doIt(args)
		} catch (e: Throwable) {
			return e
		}
		return null
	}

	val isCancellable: Boolean
		get() = true

	fun canInterruptOnCancel(): Boolean {
		return false
	}

	protected fun setResult(result: Any?) {
		this.result = result
	}

	fun <T> getResult(): T? {
		return result as T?
	}

	private fun onCancelButtonClicked(dialog: DialogInterface, which: Int) {
		cancel(canInterruptOnCancel())
		onCancelButtonClicked()
		dialog.dismiss()
	}

	protected open fun onCancelButtonClicked() {}
	override fun onActivityCreated(activity: Activity, savedInstanceState: Bundle?) {}
	override fun onActivityStarted(activity: Activity) {}
	override fun onActivityResumed(activity: Activity) {}
	override fun onActivityPaused(activity: Activity) {}
	override fun onActivityStopped(activity: Activity) {
		cancel(true)
	}

	override fun onActivitySaveInstanceState(activity: Activity, outState: Bundle) {}
	override fun onActivityDestroyed(activity: Activity) {}
}
